python/cudf/cudf/core/udf/function.cu

```diff
@@ -0,0 +1,817 @@
+  int ITEMS_PER_THREAD = (size + tb_size - 1) / tb_size;
+  int64_t local_sum = 0;
+
+  __shared__ int64_t sum;
+
+  if (tid == 0)
+    sum = 0;
+
+  __syncthreads();
+
+  // Calculate local sum for each thread
+  #pragma unroll
+  for (int ITEM = 0; ITEM < ITEMS_PER_THREAD; ITEM++) {
+      if (tid + (ITEM * tb_size) < size) {
+          int64_t load = data[tid + ITEM * tb_size];
+          local_sum += load;
+      }
+  }
+
+  atomicAdd((unsigned long long*) &sum, (unsigned long long) local_sum);
+
+  __syncthreads();
+
+  *numba_return_value = sum;
+
+  return 0;
+}
+
+extern "C" __device__ int BlockSum_float64(double *numba_return_value, double *data, int64_t size) {
+  int tid = threadIdx.x; int tb_size = blockDim.x;
+  // Calculate how many elements each thread is working on
+  int ITEMS_PER_THREAD = (size + tb_size - 1) / tb_size;
+  double local_sum = 0;
+
+  __shared__ double sum;
+
+  if (tid == 0)
+    sum = 0;
+
+  __syncthreads();
+
+  // Calculate local sum for each thread
+  #pragma unroll
+  for (int ITEM = 0; ITEM < ITEMS_PER_THREAD; ITEM++) {
+      if (tid + (ITEM * tb_size) < size) {
+          double load = data[tid + ITEM * tb_size];
+          local_sum += load;
+      }
+  }
+
+  atomicAdd(&sum, local_sum);
+
+  __syncthreads();
+
+  *numba_return_value = sum;
+
+  return 0;
+}
+
+
+extern "C" __device__ int BlockMean_int64(double *numba_return_value, int64_t *data, int64_t size) {
+  int tid = threadIdx.x; int tb_size = blockDim.x;
+  // Calculate how many elements each thread is working on
+  int ITEMS_PER_THREAD = (size + tb_size - 1) / tb_size;
+  int64_t local_sum = 0;
+  double mean;
+
+  __shared__ int64_t sum;
+
+  if (tid == 0)
+    sum = 0;
+
+  __syncthreads();
+
+  // Calculate local sum for each thread
+  #pragma unroll
+  for (int ITEM = 0; ITEM < ITEMS_PER_THREAD; ITEM++) {
+      if (tid + (ITEM * tb_size) < size) {
+          int64_t load = data[tid + ITEM * tb_size];
+          local_sum += load;
+      }
+  }
+
+  atomicAdd((unsigned long long*) &sum, (unsigned long long) local_sum);
+
+  __syncthreads();
+
+  mean = sum * 1.0 / size;
+
+  *numba_return_value = mean;
+
+  return 0;
+
+}
+
+extern "C" __device__ int BlockMean_float64(double *numba_return_value, double *data, int64_t size) {
+  int tid = threadIdx.x; int tb_size = blockDim.x;
+  // Calculate how many elements each thread is working on
+  int ITEMS_PER_THREAD = (size + tb_size - 1) / tb_size;
+  double local_sum = 0;
+  double mean;
+
+  __shared__ double sum;
+
+  if (tid == 0)
+    sum = 0;
+
+  __syncthreads();
+
+  // Calculate local sum for each thread
+  #pragma unroll
+  for (int ITEM = 0; ITEM < ITEMS_PER_THREAD; ITEM++) {
+      if (tid + (ITEM * tb_size) < size) {
+          double load = data[tid + ITEM * tb_size];
+          local_sum += load;
+      }
+  }
+
+  atomicAdd(&sum, local_sum);
+
+  __syncthreads();
+
+  mean = sum * 1.0 / size;
+
+  *numba_return_value = mean;
```

Comment:
Is the mean value supposed to be computed and written to the return pointer by all threads?

Reply:
Yes, this is how it currently works in the C++ layer. Later on, only thread 0 will write the mean to the output array though.